packages/build-tools/src/steps/functions/cache.ts

```diff
@@ -0,0 +1,104 @@
+import { BuildJob } from '@expo/eas-build-job';
+import {
+  BuildFunction,
+  BuildStepInput,
+  BuildStepInputValueTypeName,
+  BuildStepOutput,
+} from '@expo/steps';
+
+import { CustomBuildContext } from '../../customBuildContext';
+
+export function createRestoreCacheBuildFunction(ctx: CustomBuildContext): BuildFunction {
+  return new BuildFunction({
+    namespace: 'eas',
+    id: 'restore-cache',
+    name: 'Restore Cache',
+    inputProviders: [
+      BuildStepInput.createProvider({
+        id: 'key',
+        required: true,
+        allowedValueTypeName: BuildStepInputValueTypeName.STRING,
+      }),
+      BuildStepInput.createProvider({
+        id: 'paths',
+        required: true,
+        defaultValue: [],
+        allowedValueTypeName: BuildStepInputValueTypeName.JSON,
+      }),
+    ],
+    outputProviders: [
+      BuildStepOutput.createProvider({
+        id: 'cache_key',
+        required: true,
+      }),
+      BuildStepOutput.createProvider({
+        id: 'cache_paths',
+        required: false,
+      }),
+    ],
+    fn: async (stepsCtx, { inputs, outputs }) => {
+      const cacheManager = ctx.runtimeApi.cacheManager;
+
+      const key = inputs.key.value as string;
+      const paths = inputs.paths.value as [string];
+      outputs.cache_key.set(key);
+      if (paths.length > 0) {
+        outputs.cache_paths.set(JSON.stringify(paths));
+      }
+
+      if (!cacheManager) {
+        stepsCtx.logger.warn('Cache manager is not available, skipping...');
+        return;
+      }
+
+      const job = stepsCtx.global.staticContext.job as BuildJob;
```

Comment:
If the function only makes sense for a `BuildJob` let's define argument to this function like https://github.com/expo/eas-build/blob/e15bbd61ccd025b2846d480e0cf454db962b934d/packages/build-tools/src/steps/functions/resolveBuildConfig.ts#L9-L11 so types require a `BuildJob` and we don't need to cast.

However, I think the bigger thing to update here is that I think this doesn't take into consideration updates to https://github.com/expo/turtle-v2/pull/1715, e.g. that cache download URLs are to be requested dynamically from the launcher?

Reply:
- call to `restoreCache` depends on `job.cache` existence
- `restoreCache` receives `job.cache`

I thought dynamic caches are completely separate from `job.cache`.